File: python/euler/bin/p0099.py
# project euler: problem 99

#   log10 base^exp = exp * (log10 base)

from collections.abc import Iterator
from math import log10
from typing import TextIO


def compute(fh: TextIO) -> str:
    def parse_data(fh: TextIO) -> Iterator[tuple[int, int]]:
        return map(
            lambda lst: (int(lst[0]), int(lst[1])),
            [line.split(',') for line in fh.read().splitlines()],
        )

    calc_result = map(lambda tpl: tpl[1] * log10(tpl[0]), parse_data(fh))

    return str(sorted(enumerate(calc_result), key=lambda x: x[1])[-1][0] + 1)


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p099_base_exp.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0099.ts
// project euler: problem 99

import { zip } from "@std/collections";
import { assetData } from "../lib/asset.ts";
import { range } from "../lib/util.ts";

export const compute = (data: string): string => {
  const parseData = (data: string): [number, number][] => {
    const splitLines = (str: string): string[] => {
      const result = str.split(/\r?\n/);
      if (result.at(-1) === "") {
        return result.slice(0, -1);
      } else {
        return result;
      }
    };

    return splitLines(data).map((x) => x.split(",")).map((
      x,
    ) => [Number(x[0]), Number(x[1])]);
  };

  const calc_result = parseData(data).map((tpl) => tpl[1] * Math.log10(tpl[0]));

  return String(
    zip(range(1, calc_result.length + 1), calc_result).sort((a, b) =>
      Number(b[1]) - Number(a[1])
    )[0][0],
  );
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p099_base_exp.txt"));
  return compute(data);
};
